### ISSUE TITLE
[Future potential issue][Far Future feature] What to do with vanilla multiplayer connection features and vanilla netcode deficiencies??

### ISSUE THREAD

**Issue by Solxanich** [tModLoader/tModLoader#1304]
### Description
From discord's collaborator channel:

Future potential issue:
"1.4 in [vanilla] MP is a lagfest" 
"I'm certain that with the addition of modded stuff its gonna be even more laggy. its gonna be pretty much impossible to play by steam" 

Far future feature reimplementation:
"The other issue is that we previously have allowed for option for vanilla clients to connect to tML servers, so I'm wary of changing vanilla and breaking that" 
"it's a "nice to have" for the future, currently, it just doesn't work one bit, due to the very first packet"

Far future potential feature:
"netcode.tml.cs and just hijack the hook"
"We can just keep the original netcode mostly untouched and then have a completely separate set of files for own version"
"it wouldn't be just faster, it would be more secure, and you know how many changes that would involve"
"most importantly, [doing this] doesn't solve any tML related issue, and [may best be handled by the vanilla devs if fixing netcode is needed]" 
we "will not do modded clients connecting to vanilla servers"  CB. Will prevent cheating by keeping this not functional.

### What does this proposal attempt to solve or improve?
This discussion notes covers possibilities as to what to do with handling how tML connects to vanilla servers, and the reverse of vanilla connecting to tML servers.
Additionally, it mentions a potential issue for lag in 1.4 multiplayer.
Further, it also suggests building a separate set of tML netcode under its own path to improve latency, stability, and security, but also suggests that it's a bit beyond the [at time of writing] scope of tML to do so.

----------------------------End of Discord Far Future feature notes-------------------------------------

**Comment by Solxanich** [comment]
Also, some additional CPU usage details from the current merged testing unoffiical beta, as CPU has high correlation to historical frame problems, keeping resolution constant and using dedicated server:
Vanilla 1.4 server CPU: 0.08 / 8
Vanilla 1.4 client CPU: ~2.00 / 8
tML 1.4 server CPU: 0.01 / 8
tML 1.4 client CPU: ~2.30 / 8

I don't foresee significant performance issues from tML alone for MP lag assuming the additive performance hit from mods in 1.3 is similar in 1.4. However, this issue could be raised in priority if later down the road it appears that's not the case

**Comment by Solxanich** [comment]
Further investigating to try to answer the question of what in 1.4 was changed or added to induce the lag part of this issue, so we can target that.
From internet searching:
I've found this correlation online to decreased performance due to multi-core lighting having been changed and the default usage of more cores is worsening performance: https://forums.terraria.org/index.php?threads/no-multicore-lighting-option-in-1-4-0-1-1-4-0-3.90616/
And this statement from Leinfor that it was removed in 1.4.0.3: <https://forums.terraria.org/index.php?threads/multicore-lighting-no-longer-available-replacement-of-this-not-working-properly-causing-the-performance-issues-reported.90046/> 
Additionally, 1.4.0.4 fixed a memory leak on servers.            
   
I don't see anything meaningful showing up in single client multiplayer performance profiling of 1.4 tML. 
Rendering and lighting continue to dominate usage at >40% and ~15% of total usage respectively.

I see two possible paths to proceed down here for further investigating: 
Path 1 is to look into the multi-core lighting/rendering claim. It's possible lag is being created by this change.
Path 2 is to try and get two multiplayer sessions with two players and three players to look at what's increasing per player in profiling
